File: Wumpus_World_Java_Shell/src/World.java
// ======================================================================
// FILE:        World.java
//
// AUTHOR:      Abdullah Younis
//
// DESCRIPTION: This file contains the world class, which is responsible
//              for everything game related.
//
// NOTES:       - Don't make changes to this file.
// ======================================================================

import java.util.Random;
import java.io.IOException;
import java.io.File;
import java.util.Scanner;

public class World
{
	// ===============================================================
	// =						Declarations
	// ===============================================================	
	
	// Tile Structure
	private class Tile
	{
		boolean pit    = false;
		boolean wumpus = false;
		boolean gold   = false;
		boolean breeze = false;
		boolean stench = false;
	}
	
	// Operation Variables
	private boolean 	debug;			// If true, displays board info after every move
	private boolean		manualAI;		// If true, alters the behavior of debug for flow purposes
	
	// Agent Variables
	private Agent 	agent;			// The agent
	private int 	score;			// The agent's score
	private boolean	goldLooted;		// True if gold was successfuly looted
	private boolean	hasArrow;		// True if the agent can shoot
	private boolean	bump;			// Bump percept flag
	private boolean	scream;			// Scream percept flag
	private int		agentDir;		// The direction the agent is facing: 0 - right, 1 - down, 2 - left, 3 - up
	private int		agentX;			// The column where the agent is located ( x-coord = col-coord )
	private int		agentY;			// The row where the agent is located ( y-coord = row-coord )

	private Agent.Action	lastAction;	// The last action the agent made
	
	// Board Variables
	private int			colDimension;	// The number of columns the game board has
	private int			rowDimension;	// The number of rows the game board has
	private Tile[][]	board;			// The game board
	
	// Random Variable
	private Random rand;
	
	// ===============================================================
	// =						Constructors
	// ===============================================================	
	
	public World ( ) throws Exception
	{
		this ( false, false, false, null );
	}
	
	public World ( boolean _debug, boolean _randomAI, boolean _manualAI ) throws Exception
	{
		this ( _debug, _randomAI, _manualAI, null );
	}
	
	public World ( boolean _debug, boolean _randomAI, boolean _manualAI, File worldFile ) throws Exception
	{
		// Operation Flags
		debug        = _debug;
		manualAI     = _manualAI;
		
		// Agent Initialization
		goldLooted   = false;
		hasArrow     = true;
		bump         = false;
		scream       = false;
		score        = 0;
		agentDir     = 0;
		agentX       = 0;
		agentY       = 0;
		lastAction   = Agent.Action.CLIMB;
		
		if ( _randomAI )
			agent = new RandomAI();
		else if ( _manualAI )
			agent = new ManualAI();
		else
			agent = new MyAI();
		
		// Board Initialization
		if ( worldFile != null )
		{
			Scanner scan = new Scanner ( worldFile );
		
			if ( !scan.hasNextInt() )
				throw new Exception();
			
			colDimension = scan.nextInt();
			
			if ( !scan.hasNextInt() )
				throw new Exception();
			
			rowDimension = scan.nextInt();
			
			board = new Tile[colDimension][rowDimension];
		
			for ( int r = 0; r < rowDimension; ++r )
				for ( int c = 0; c < colDimension; ++c )
					board[c][r] = new Tile();

			addFeatures ( scan );
		}
		else
		{
			rand = new Random ( );
			colDimension = 4;
			rowDimension = 4;
			board = new Tile[colDimension][rowDimension];
		
			for ( int r = 0; r < rowDimension; ++r )
				for ( int c = 0; c < colDimension; ++c )
					board[c][r] = new Tile();

			addFeatures ( );
		}
	}
	
	// ===============================================================
	// =					Engine Function
	// ===============================================================	
	
	public int run ( )
	{
		while ( score >= -1000 )
		{
			if ( debug || manualAI )
			{
				printWorldInfo();
				
				if ( !manualAI )
				{
					// Pause the game, only if manualAI isn't on
					// because manualAI pauses for us
					System.out.println("Press ENTER to continue...");
					try
					{
						System.in.read();
					}
					catch (IOException ioe)
					{
					}
				}
			}
			// Get the move
			lastAction = agent.getAction
			(
				board[agentX][agentY].stench,
				board[agentX][agentY].breeze,
				board[agentX][agentY].gold,
				bump,
				scream
			);

			// Make the move
			--score;
			bump   = false;
			scream = false;
			
			switch ( lastAction )
			{
				case TURN_LEFT:
					if (--agentDir < 0) agentDir = 3;
					break;
					
				case TURN_RIGHT:
					if (++agentDir > 3) agentDir = 0;
					break;
					
				case FORWARD:
					if ( agentDir == 0 && agentX+1 < colDimension )
						++agentX;
					else if ( agentDir == 1 && agentY-1 >= 0 )
						--agentY;
					else if ( agentDir == 2 && agentX-1 >= 0 )
						--agentX;
					else if ( agentDir == 3 && agentY+1 < rowDimension )
						++agentY;
					else
						bump = true;
					
					if ( board[agentX][agentY].pit || board[agentX][agentY].wumpus )
					{
						score -= 1000;
						if (debug) printWorldInfo();
						return score;
					}
					break;
				
				case SHOOT:
					if ( hasArrow )
					{
						hasArrow = false;
						score -= 10;
						if ( agentDir == 0 )
						{
							for ( int x = agentX; x < colDimension; ++x )
								if ( board[x][agentY].wumpus )
								{
									board[x][agentY].wumpus = false;
									board[x][agentY].stench = true;
									scream = true;
								}
						}
						else if ( agentDir == 1 )
						{
							for ( int y = agentY; y >= 0; --y )
								if ( board[agentX][y].wumpus )
								{
									board[agentX][y].wumpus = false;
									board[agentX][y].stench = true;
									scream = true;
								}
						}
						else if ( agentDir == 2 )
						{
							for ( int x = agentX; x >= 0; --x )
								if ( board[x][agentY].wumpus )
								{
									board[x][agentY].wumpus = false;
									board[x][agentY].stench = true;
									scream = true;
								}
						}
						else if ( agentDir == 3 )
						{
							for ( int y = agentY; y < rowDimension; ++y )
								if ( board[agentX][y].wumpus )
								{
									board[agentX][y].wumpus = false;
									board[agentX][y].stench = true;
									scream = true;
								}
						}
					}
					break;
					
				case GRAB:
					if ( board[agentX][agentY].gold )
					{
						board[agentX][agentY].gold = false;
						goldLooted = true;
					}
					break;
					
				case CLIMB:
					if ( agentX == 0 && agentY == 0 )
					{
						if ( goldLooted )
							score += 1000;
						if (debug) printWorldInfo();
						return score;
					}
					break;
			}
		}
		return score;
	}
	// ===============================================================
	// =				World Generation Functions
	// ===============================================================
	
	private void addFeatures ( )
	{		
		// Generate pits
		for ( int r = 0; r < rowDimension; ++r )
			for ( int c = 0; c < colDimension; ++c )
				if ( (c != 0 || r != 0) && randomInt(10) < 2 )
					addPit ( c, r );
		
		// Generate wumpus
		int wc = randomInt(colDimension);
		int wr = randomInt(rowDimension);
		
		while ( wc == 0 && wr == 0 )
		{
			wc = randomInt(colDimension);
			wr = randomInt(rowDimension);
		}
		
		addWumpus ( wc, wr );
		
		// Generate gold
		int gc = randomInt(colDimension);
		int gr = randomInt(rowDimension);
			
		while ( gc == 0 && gr == 0 )
		{
			gc = randomInt(colDimension);
			gr = randomInt(rowDimension);
		}
		
		addGold ( gc, gr );
	}
	
	private void addFeatures ( Scanner scan ) throws Exception
	{
		int c, r;
		
		// Add the Wumpus
		if ( !scan.hasNextInt() )
			throw new Exception();
		
		c = scan.nextInt();
		
		if ( !scan.hasNextInt() )
			throw new Exception();
		
		r = scan.nextInt();
		
		addWumpus ( c, r );
		
		// Add the Gold
		
		if ( !scan.hasNextInt() )
			throw new Exception();
		
		c = scan.nextInt();
		
		if ( !scan.hasNextInt() )
			throw new Exception();
		
		r = scan.nextInt();
		
		addGold ( c, r );
		
		// Add the Pits
		
		if ( !scan.hasNextInt() )
			throw new Exception();
		
		int numOfPits = scan.nextInt();
		
		while ( numOfPits > 0 && scan.hasNextLine() )
		{
			--numOfPits;
		
			if ( !scan.hasNextInt() )
				throw new Exception();
			
			c = scan.nextInt();
		
			if ( !scan.hasNextInt() )
				throw new Exception();
			
			r = scan.nextInt();
			
			addPit ( c, r );
		}

		scan.close();
	}
	
	private void addPit ( int c, int r )
	{
		if ( isInBounds(c, r) )
		{
			board[c][r].pit = true;
			addBreeze ( c+1, r );
			addBreeze ( c-1, r );
			addBreeze ( c, r+1 );
			addBreeze ( c, r-1 );
		}
	}
	
	private void addWumpus ( int c, int r )
	{
		if ( isInBounds(c, r) )
		{
			board[c][r].wumpus = true;
			addStench ( c+1, r );
			addStench ( c-1, r );
			addStench ( c, r+1 );
			addStench ( c, r-1 );
		}
	}
	
	private void addGold ( int c, int r )
	{
		if ( isInBounds(c, r) )
			board[c][r].gold = true;
	}
	
	private void addStench ( int c, int r )
	{
		if ( isInBounds(c, r) )
			board[c][r].stench = true;
	}
	
	private void addBreeze ( int c, int r )
	{
		if ( isInBounds(c, r) )
			board[c][r].breeze = true;
	}
	
	private boolean isInBounds ( int c, int r )
	{
		return ( c < colDimension && r < rowDimension && c >= 0  && r >= 0);
	}
	
	// ===============================================================
	// =				World Printing Functions
	// ===============================================================
	
	private void printWorldInfo ( )
	{
		printBoardInfo();
		printAgentInfo();
	}
	
	private void printBoardInfo ( )
	{
		for ( int r = rowDimension-1; r >= 0; --r )
		{
			for ( int c = 0; c < colDimension; ++c )
				printTileInfo ( c, r );
			System.out.println("");
			System.out.println("");
		}
	}

	private void printTileInfo ( int c, int r )
	{
		StringBuilder tileString = new StringBuilder();
		
		if (board[c][r].pit)    tileString.append("P");
		if (board[c][r].wumpus) tileString.append("W");
		if (board[c][r].gold)   tileString.append("G");
		if (board[c][r].breeze) tileString.append("B");
		if (board[c][r].stench) tileString.append("S");
		
		if ( agentX == c && agentY == r )
			tileString.append("@");
		
		tileString.append(".");
		
		System.out.printf("%8s", tileString.toString());
	}
	
	private void printAgentInfo ( )
	{
		System.out.println("Score: "   + score);
		System.out.println("AgentX: "  + agentX);
		System.out.println("AgentY: "  + agentY);
		printDirectionInfo();
		printActionInfo();
		printPerceptInfo();
	}
	
	private void printDirectionInfo ( )
	{
		switch (agentDir)
		{
			case 0:
				System.out.println("AgentDir: Right");
				break;
				
			case 1:
				System.out.println("AgentDir: Down");
				break;
				
			case 2:
				System.out.println("AgentDir: Left");
				break;
				
			case 3:
				System.out.println("AgentDir: Up");
				break;
				
			default:
				System.out.println("AgentDir: Invalid");
		}
	}
	
	private void printActionInfo ( )
	{
		switch (lastAction)
		{
			case TURN_LEFT:
				System.out.println("Last Action: Turned Left");
				break;
				
			case TURN_RIGHT:
				System.out.println("Last Action: Turned Right");
				break;
				
			case FORWARD:
				System.out.println("Last Action: Moved Forward");
				break;
				
			case SHOOT:
				System.out.println("Last Action: Shot the arrow");
				break;
				
			case GRAB:
				System.out.println("Last Action: Grabbed");
				break;
				
			case CLIMB:
				System.out.println("Last Action: Climbed");
				break;
				
			default:
				System.out.println("Last Action: Invalid");
		}
	}
	
	private void printPerceptInfo ( )
	{
		StringBuilder perceptString = new StringBuilder("Percepts: ");
		
		if (board[agentX][agentY].stench) perceptString.append("Stench, ");
		if (board[agentX][agentY].breeze) perceptString.append("Breeze, ");
		if (board[agentX][agentY].gold)   perceptString.append("Glitter, ");
		if (bump)                         perceptString.append("Bump, ");
		if (scream)                       perceptString.append("Scream");
		
		if ( perceptString.charAt(perceptString.length()-1) == ' '
				&& perceptString.charAt(perceptString.length()-2) == ',' )
		{
			perceptString.deleteCharAt(perceptString.length()-1);
			perceptString.deleteCharAt(perceptString.length()-1);
		}
		
		System.out.println(perceptString.toString());
	}
	
	// ===============================================================
	// =					Helper Functions
	// ===============================================================
	
	private int randomInt ( int limit )
	{
		return rand.nextInt(limit);
	}
}

File: Wumpus_World_Python_Shell/src/World.py
# ======================================================================
# FILE:        World.py
#
# AUTHOR:      Abdullah Younis
#
# DESCRIPTION: This file contains the world class, which is responsible
#              for everything game related.
#
# NOTES:       - Don't make changes to this file.
# ======================================================================

from Agent import Agent
from ManualAI import ManualAI
from RandomAI import RandomAI
from MyAI import MyAI
import random

class World():
    
    # Tile Structure
    class __Tile:
        pit    = False;
        wumpus = False;
        gold   = False;
        breeze = False;
        stench = False;
    
    # ===============================================================
    # =                 Constructor
    # ===============================================================   
    
    def __init__ ( self, debug = False, randomAI = False, manualAI = False, file = None ):
        # Operation Flags
        self.__debug        = debug
        self.__manualAI      = manualAI
        
        # Agent Initialization
        self.__goldLooted   = False
        self.__hasArrow     = True
        self.__bump         = False
        self.__scream       = False
        self.__score        = 0
        self.__agentDir     = 0
        self.__agentX       = 0
        self.__agentY       = 0
        self.__lastAction   = Agent.Action.CLIMB
        
        if randomAI:
            self.__agent = RandomAI()
        elif manualAI:
            self.__agent = ManualAI()
        else:
            self.__agent = MyAI()
            
        if file != None:
            self.__colDimension, self.__rowDimension = [int(x) for x in next(file).split()]
            self.__board = [[self.__Tile() for j in range(self.__rowDimension)] for i in range(self.__colDimension)]
            self.__addFeatures(file)
        else:
            self.__colDimension = 4
            self.__rowDimension = 4
            self.__board = [[self.__Tile() for j in range(self.__colDimension)] for i in range(self.__rowDimension)]
            self.__addFeatures()
    
    # ===============================================================
    # =                 Engine Function
    # ===============================================================   
    
    def run ( self ):
        while self.__score >= -1000:
            if self.__debug or self.__manualAI:
                self.__printWorldInfo()
                
                if not self.__manualAI:
                    # Pause the game, only if manualAI isn't on
                    # because manualAI pauses for us
                    input("Press ENTER to continue...")
                        
            # Get the move
            self.__lastAction = self.__agent.getAction (
														self.__board[self.__agentX][self.__agentY].stench,
														self.__board[self.__agentX][self.__agentY].breeze,
														self.__board[self.__agentX][self.__agentY].gold,
														self.__bump,
														self.__scream
													   )

            # Make the move
            self.__score -= 1;
            self.__bump   = False;
            self.__scream = False;
            
            if self.__lastAction == Agent.Action.TURN_LEFT:
                self.__agentDir -= 1
                if (self.__agentDir < 0):
                    self.__agentDir = 3
                    
            elif self.__lastAction == Agent.Action.TURN_RIGHT:
                self.__agentDir += 1
                if self.__agentDir > 3:
                    self.__agentDir = 0
                    
            elif self.__lastAction == Agent.Action.FORWARD:
                if self.__agentDir == 0 and self.__agentX+1 < self.__colDimension:
                    self.__agentX += 1
                elif self.__agentDir == 1 and self.__agentY-1 >= 0:
                    self.__agentY -= 1
                elif self.__agentDir == 2 and self.__agentX-1 >= 0:
                    self.__agentX -= 1
                elif self.__agentDir == 3 and self.__agentY+1 < self.__rowDimension:
                    self.__agentY += 1
                else:
                    self.__bump = True
                    
                if self.__board[self.__agentX][self.__agentY].pit or self.__board[self.__agentX][self.__agentY].wumpus:
                    self.__score -= 1000
                    if self.__debug:
                        self.__printWorldInfo()
                    return self.__score
                
            elif self.__lastAction == Agent.Action.SHOOT:
            
                if self.__hasArrow:
                    self.__hasArrow = False
                    self.__score -= 10
                    
                    if self.__agentDir == 0:
                        for x in range (self.__agentX, self.__colDimension):
                                if self.__board[x][self.__agentY].wumpus:
                                    self.__board[x][self.__agentY].wumpus = False
                                    self.__board[x][self.__agentY].stench = True
                                    self.__scream = True
                    
                    elif self.__agentDir == 1:
                        for y in range (self.__agentY, -1, -1):
                            if self.__board[self.__agentX][y].wumpus:
                                self.__board[self.__agentX][y].wumpus = False
                                self.__board[self.__agentX][y].stench = True
                                self.__scream = True
                    
                    elif self.__agentDir == 2:
                        for x in range (self.__agentX, -1, -1):
                            if self.__board[x][self.__agentY].wumpus:
                                self.__board[x][self.__agentY].wumpus = False
                                self.__board[x][self.__agentY].stench = True
                                self.__scream = True

                    elif self.__agentDir == 3:
                        for y in range (self.__agentY, self.__rowDimension):
                            if self.__board[self.__agentX][y].wumpus:
                                self.__board[self.__agentX][y].wumpus = False
                                self.__board[self.__agentX][y].stench = True
                                self.__scream = True
                    
            elif self.__lastAction == Agent.Action.GRAB:
                if self.__board[self.__agentX][self.__agentY].gold:
                    self.__board[self.__agentX][self.__agentY].gold = False
                    self.__goldLooted = True
                    
            elif self.__lastAction == Agent.Action.CLIMB:
                if self.__agentX == 0 and self.__agentY == 0:
                    if self.__goldLooted:
                        self.__score += 1000
                    if (self.__debug):
                        self.__printWorldInfo()
                    return self.__score;
        return self.__score
        
    # ===============================================================
    # =             World Generation Functions
    # ===============================================================
    
    def __addFeatures ( self, file = None ):
        if file == None:
            # Generate pits
            for r in range (self.__rowDimension):
                for c in range (self.__colDimension):
                    if (c != 0 or r != 0) and self.__randomInt(10) < 2:
                        self.__addPit ( c, r )
            
            # Generate wumpus
            wc = self.__randomInt(self.__colDimension)
            wr = self.__randomInt(self.__rowDimension)
            
            while wc == 0 and wr == 0:
                wc = self.__randomInt(self.__colDimension)
                wr = self.__randomInt(self.__rowDimension)
                
            self.__addWumpus ( wc, wr );
            
            # Generate gold
            gc = self.__randomInt(self.__colDimension)
            gr = self.__randomInt(self.__rowDimension)
                
            while gc == 0 and gr == 0:
                gc = self.__randomInt(self.__colDimension)
                gr = self.__randomInt(self.__rowDimension)
            
            self.__addGold ( gc, gr )
            
        else:
            # Add the Wumpus
            c, r = [int(x) for x in next(file).split()]
            self.__addWumpus ( c, r )
            
            # Add the Gold
            c, r = [int(x) for x in next(file).split()]
            self.__addGold ( c, r )
            
            # Add the Pits
            numOfPits = int(next(file))
            
            while numOfPits > 0:
                numOfPits -= 1
                c, r = [int(x) for x in next(file).split()]
                self.__addPit ( c, r )
                
            file.close()
    
    def __addPit ( self, c, r ):
        if self.__isInBounds(c, r):
            self.__board[c][r].pit = True
            self.__addBreeze ( c+1, r )
            self.__addBreeze ( c-1, r )
            self.__addBreeze ( c, r+1 )
            self.__addBreeze ( c, r-1 )
    
    def __addWumpus ( self, c, r ):
        if self.__isInBounds(c, r):
            self.__board[c][r].wumpus = True
            self.__addStench ( c+1, r )
            self.__addStench ( c-1, r )
            self.__addStench ( c, r+1 )
            self.__addStench ( c, r-1 )
    
    def __addGold ( self, c, r ):
        if self.__isInBounds(c, r):
            self.__board[c][r].gold = True
    
    def __addStench ( self, c, r ):
        if self.__isInBounds(c, r):
            self.__board[c][r].stench = True
    
    def __addBreeze ( self, c, r ):
        if self.__isInBounds(c, r):
            self.__board[c][r].breeze = True
    
    def __isInBounds ( self, c, r ):
        return c < self.__colDimension and r < self.__rowDimension and c >= 0 and r >= 0
    
    # ===============================================================
    # =             World Printing Functions
    # ===============================================================
    
    def __printWorldInfo ( self ):
        self.__printBoardInfo()
        self.__printAgentInfo()
    
    def __printBoardInfo ( self ):
        for r in range (self.__rowDimension-1, -1, -1):
            for c in range (self.__colDimension):
                self.__printTileInfo ( c, r )
            print("")
            print("")

    def __printTileInfo ( self, c, r ):
        tileString = ""
        
        if self.__board[c][r].pit:    tileString += "P"
        if self.__board[c][r].wumpus: tileString += "W"
        if self.__board[c][r].gold:   tileString += "G"
        if self.__board[c][r].breeze: tileString += "B"
        if self.__board[c][r].stench: tileString += "S"
        
        if self.__agentX == c and self.__agentY == r:
            tileString += "@"
        
        tileString += "."
        
        print(tileString.rjust(8), end="")
    
    def __printAgentInfo ( self ):
        print ( "Score: "   + str(self.__score) )
        print ( "AgentX: "  + str(self.__agentX) )
        print ( "AgentY: "  + str(self.__agentY) )
        self.__printDirectionInfo()
        self.__printActionInfo()
        self.__printPerceptInfo()
    
    def __printDirectionInfo ( self ):
        if self.__agentDir == 0:
            print ( "AgentDir: Right" )
            
        elif self.__agentDir == 1:
            print ( "AgentDir: Down" )
            
        elif self.__agentDir == 2:
            print ( "AgentDir: Left" )
            
        elif self.__agentDir == 3:
            print ( "AgentDir: Up" )
            
        else:
            print ( "AgentDir: Invalid" )
    
    def __printActionInfo ( self ):
        if self.__lastAction == Agent.Action.TURN_LEFT:
            print ( "Last Action: Turned Left" )

        elif self.__lastAction == Agent.Action.TURN_RIGHT:
            print ( "Last Action: Turned Right")

        elif self.__lastAction == Agent.Action.FORWARD:
            print ( "Last Action: Moved Forward")

        elif self.__lastAction == Agent.Action.SHOOT:
            print ( "Last Action: Shot the Arrow")

        elif self.__lastAction == Agent.Action.GRAB:
            print ( "Last Action: Grabbed")

        elif self.__lastAction == Agent.Action.CLIMB:
            print ( "Last Action: Climbed")

        else:
            print ( "Last Action: Invalid")

    def __printPerceptInfo ( self ):
        perceptString = "Percepts: "
        
        if self.__board[self.__agentX][self.__agentY].stench: perceptString += "Stench, "
        if self.__board[self.__agentX][self.__agentY].breeze: perceptString += "Breeze, "
        if self.__board[self.__agentX][self.__agentY].gold:   perceptString += "Glitter, "
        if self.__bump:                         perceptString += "Bump, "
        if self.__scream:                       perceptString += "Scream"
        
        if perceptString[-1] == ' 'and perceptString[-2] == ',':
            perceptString = perceptString[:-2]
        
        print(perceptString)
    
    # ===============================================================
    # =                 Helper Functions
    # ===============================================================
    
    def __randomInt ( self, limit ):
        return random.randrange(limit)
